components/tidb_query/src/rpn_expr/impl_cast.rs

```diff
@@ -226,55 +344,342 @@ pub fn cast_any_as_decimal<From: Evaluable + ConvertTo<Decimal>>(
+            }
+        }
+    }
+}
+
+#[rpn_fn(capture = [ctx, extra])]
+#[inline]
+fn cast_string_as_int_or_uint(
+    ctx: &mut EvalContext,
+    extra: &RpnFnCallExtra<'_>,
+    val: &Option<Bytes>,
+) -> Result<Option<i64>> {
+    match val {
+        None => Ok(None),
+        Some(val) => {
+            // TODO, in TiDB',s if `b.args[0].GetType().Hybrid()`,
+            //  then it will return res from EvalInt() directly.
+            let is_unsigned = is_field_type_unsigned(extra.ret_field_type);
+            let val = get_valid_utf8_prefix(ctx, val.as_slice())?;
+            let val = val.trim();
+            let neg = val.starts_with('-');
+            if !neg {
+                let r: crate::codec::error::Result<u64> = val.as_bytes().convert(ctx);
+                match r {
+                    Ok(x) => {
+                        if !is_unsigned && x > std::i64::MAX as u64 {
+                            ctx.warnings
+                                .append_warning(Error::cast_as_signed_overflow())
+                        }
+                        Ok(Some(x as i64))
+                    }
+                    Err(e) => handle_overflow_for_cast_string_as_int(ctx, e, false, val).map(Some),
+                }
+            } else if in_union(extra.implicit_args) && is_unsigned {
+                Ok(Some(0))
+            } else {
+                let r: crate::codec::error::Result<i64> = val.as_bytes().convert(ctx);
+                match r {
+                    Ok(x) => {
+                        if is_unsigned {
+                            ctx.warnings
+                                .append_warning(Error::cast_neg_int_as_unsigned());
+                        }
+                        Ok(Some(x))
+                    }
+                    Err(e) => handle_overflow_for_cast_string_as_int(ctx, e, true, val).map(Some),
+                }
+            }
+        }
+    }
+}
+
+// TODO, TiDB's this func can return err and res at the same time, however, we can't,
+//  so it may be some inconsistency between this func and TiDB's
+fn handle_overflow_for_cast_string_as_int(
+    ctx: &mut EvalContext,
+    err: Error,
+    is_neg: bool,
+    origin_str: &str,
+) -> Result<i64> {
+    if ctx.cfg.flag.contains(Flag::IN_INSERT_STMT) && err.is_overflow() {
+        // TODO, here our `handle_overflow_err` is not same as TiDB's HandleOverflow,
+        //  the latter will return `err` if OverflowAsWarning, but we will return `truncated_wrong_val`.
+        ctx.handle_overflow_err(Error::truncated_wrong_val("INTEGER", origin_str))?;
+        if is_neg {
+            Ok(std::i64::MIN)
+        } else {
+            Ok(std::u64::MAX as i64)
+        }
+    } else {
+        Err(err.into())
+    }
+}
+
 #[rpn_fn(capture = [ctx])]
 #[inline]
-fn cast_any_as_any<From: ConvertTo<To> + Evaluable, To: Evaluable>(
+fn cast_decimal_as_int(ctx: &mut EvalContext, val: &Option<Decimal>) -> Result<Option<i64>> {
+    match val {
```

Comment:
This logic is the same as `ConvertTo<i64> for Decimal`,  prefer to use `cast_any_as_any`.